src/promise.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Returns a Deferred struct, which holds a pending promise and its associated
+ * resolve and reject functions.
+ *
+ * @template T
+ */
+export class Deferred {
+  constructor() {
+    /**
+     * @const {!Promise<T>}
+     */
+    this.promise = new /*OK*/Promise((resolve, reject) => {
+      /**
+       * @const {function((TYPE|IThenable<TYPE>|Thenable)=)}
```

Comment:
`TYPE` seems to be tripping up the closure compiler.

https://travis-ci.org/ampproject/amphtml/jobs/375151957#L2181

https://github.com/google/closure-compiler/wiki/Bad-Type-Annotation